File: August 2025/09.py
class Solution:
    def getLongestPrefix(self, s: str) -> int:
        n = len(s)
        if n <= 1:
            return -1

        # Build Z-array: z[i] = longest substring starting at i
        # that matches a prefix of s
        z = [0] * n
        l = r = 0
        for i in range(1, n):
            if i <= r:
                # mirror within [l..r]
                z[i] = min(r - i + 1, z[i - l])
            # extend match past r
            while i + z[i] < n and s[z[i]] == s[i + z[i]]:
                z[i] += 1
            if i + z[i] - 1 > r:
                l, r = i, i + z[i] - 1

        # Scan from longest possible proper prefix downwards
        for length in range(n - 1, 0, -1):
            # if z[length] covers the rest of the string,
            # repeating the prefix of this length will match s
            if z[length] >= n - length:
                return length

        return -1

File: August 2025/09.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
  public:
    int getLongestPrefix(string &s) {
        int n = s.length();
        if (n == 0) 
            return -1;

        // Build LPS (longest proper prefix which is also suffix) array
        vector<int> lps(n, 0);
        int len = 0;   // length of the previous longest prefix suffix
        int i = 1;

        while (i < n) {
            if (s[i] == s[len]) {
                len++;
                lps[i] = len;
                i++;
            } else {
                if (len != 0) {
                    // fallback in the prefix table
                    len = lps[len - 1];
                } else {
                    lps[i] = 0;
                    i++;
                }
            }
        }

        int longest = lps[n - 1];
        // if no proper prefix-suffix or not periodic, return -1
        if (longest == 0) 
            return -1;

        // Check periodicity: total length divisible by the smallest repeating unit
        int unitLen = n - longest;
        if (n % unitLen == 0)
            return longest;
        else
            return -1;
    }
};
